### ISSUE TITLE
std::transform: candidate template ignored: couldn't infer template argument '_BinaryOperation'

### ISSUE THREAD

**Issue by aleneum** [iclcv/icl#3]
There is an issue for `std::transform` in combination with `pow` which cannot be resolved on OSX (works fine for Linux).

```
#include <cmath>
#include <vector>
#include <algorithm>

int main() {
    std::vector<float> a;
    std::vector<float> b;
    std::vector<float> dst;
    std::transform(a.begin(), a.end(), b.begin(), dst.begin(), ::pow);
    return 0;
}
```

Builds on Linux but not on OSX Sierra

**Comment by aleneum** [comment]
Posted a question on [SO](https://stackoverflow.com/questions/46567255/compiling-stdtransforms-with-argument-atan2-pow-fails-with-clang-but-works-wi)

**Comment by aleneum** [comment]
Okay, this one is for @celbrech I guess. It's just 5 year old code, so no big deal:

[context](https://github.com/iclcv/icl/blob/master/ICLMath/src/ICLMath/DynMatrixUtils.cpp#L506)

Which `::atan2` do you want to use? With `clang` this resolves to `std::atan2` which requires to specify which version of the function you need (e.g. float or double). GCC does **not** resolve it to `std::atan2` cause if you specify `std:atan2` directly, gcc also fails with the same error:

```
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:507:5: error: no matching function for call to 'transform'
    INSTANTIATE_DYN_MATRIX_MATH_OP(arctan2, ::atan2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:493:7: note: expanded from macro 'INSTANTIATE_DYN_MATRIX_MATH_OP'
      std::transform(a.begin(),a.end(),b.begin(),dst.begin(),func);       \
      ^~~~~~~~~~~~~~
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:507:5: note: in instantiation of function template specialization 'icl::math::matrix_arctan2<float>' requested
      here
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:496:44: note: expanded from macro 'INSTANTIATE_DYN_MATRIX_MATH_OP'
    template ICLMath_API DynMatrix<float> &matrix_##op(const DynMatrix<float>&, const DynMatrix<float>&, DynMatrix<float>&)     \
                                           ^
<scratch space>:74:1: note: expanded from here
matrix_arctan2
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:1932:1: note: candidate template ignored: couldn't infer template argument
      '_BinaryOperation'
transform(_InputIterator1 __first1, _InputIterator1 __last1, _InputIterator2 __first2,
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:1922:1: note: candidate function template not viable: requires 4 arguments,
      but 5 were provided
transform(_InputIterator __first, _InputIterator __last, _OutputIterator __result, _UnaryOperation __op)
^
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:507:5: error: no matching function for call to 'transform'
    INSTANTIATE_DYN_MATRIX_MATH_OP(arctan2, ::atan2)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:493:7: note: expanded from macro 'INSTANTIATE_DYN_MATRIX_MATH_OP'
      std::transform(a.begin(),a.end(),b.begin(),dst.begin(),func);       \
      ^~~~~~~~~~~~~~
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:507:5: note: in instantiation of function template specialization 'icl::math::matrix_arctan2<double>' requested
      here
/Users/alneuman/Workspace/icl/ICLMath/src/ICLMath/DynMatrixUtils.cpp:498:45: note: expanded from macro 'INSTANTIATE_DYN_MATRIX_MATH_OP'
    template ICLMath_API DynMatrix<double> &matrix_##op(const DynMatrix<double>&, const DynMatrix<double>&, DynMatrix<double>&) \
                                            ^
<scratch space>:74:1: note: expanded from here
matrix_arctan2
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:1932:1: note: candidate template ignored: couldn't infer template argument
      '_BinaryOperation'
transform(_InputIterator1 __first1, _InputIterator1 __last1, _InputIterator2 __first2,
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:1922:1: note: candidate function template not viable: requires 4 arguments,
      but 5 were provided
transform(_InputIterator __first, _InputIterator __last, _OutputIterator __result, _UnaryOperation __op)
^
2 errors generated.
```

tl;dr: There seem to be a `atan2` in global namespace of `gcc` that `clang` does not have in recent versions.

**Comment by aleneum** [comment]
Okay, I guess the GCC uses the *builtin* version `__builtin_atan2` ([ref](https://man.cx/atan2)) and clang does not because clang builtins have to be called directly. I could specify the signature (double(*)(double, double)) which should comply the builtin gcc version as well as the std version clang tries to use.

**Comment by celbrech** [comment]
This (can easily be fixed with a small wrapper function:
.. I know its ugly, for for now it seems to work fine 
Alternatively, a lambda expression would do the job (see lower example)

#include <vector>
#include <algorithm>
#include <cmath>

inline float pow_wrapper(float b, float e) { return ::pow(b,e); }
int main() {
    std::vector<float> a;
    std::vector<float> b;
    std::vector<float> dst;
    std::transform(a.begin(), a.end(), b.begin(), dst.begin(), pow_wrapper);
    return 0;
}

#include <vector>
#include <algorithm>
#include <cmath>

inline float pow_wrapper(float b, float e) { return ::pow(b,e); }
int main() {
    std::vector<float> a;
    std::vector<float> b;
    std::vector<float> dst;
    std::transform(a.begin(), a.end(), b.begin(), dst.begin(),
                   [](float b, float e) -> float { return ::pow(b,e); });
    return 0;
}

**Comment by aleneum** [comment]
@celbrech: I fixed this with a `static_cast` in case clang is used. I am currently rewriting the merge request #4 into two new requests where one targets OS related things and the other clang stuff.
